File: cmd/synth/cmd_landscape.go
package main

import (
	"fmt"
	"strings"

	"gitlab.com/nomograph/synthesist/internal/store"
)

func cmdStakeholder(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth stakeholder <add|list> ...")
	}
	switch args[0] {
	case "add":
		return cmdStakeholderAdd(args[1:])
	case "list":
		return cmdStakeholderList(args[1:])
	default:
		return fmt.Errorf("unknown stakeholder subcommand: %s", args[0])
	}
}

func cmdStakeholderAdd(args []string) error {
	if len(args) < 2 {
		return fmt.Errorf("usage: synth stakeholder add <tree> <id> --context 'role' [--name 'Full Name'] [--orgs 'org1,org2']")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree := args[0]
	id := args[1]

	var context, name string
	var orgs []string
	for i := 2; i < len(args)-1; i += 2 {
		switch args[i] {
		case "--context":
			context = args[i+1]
		case "--name":
			name = args[i+1]
		case "--orgs":
			orgs = strings.Split(args[i+1], ",")
		}
	}
	if context == "" {
		return fmt.Errorf("--context is required")
	}

	var namePtr *string
	if name != "" {
		namePtr = &name
	}

	_, err = s.DB.Exec("INSERT IGNORE INTO stakeholders (tree, id, name, context) VALUES (?, ?, ?, ?)",
		tree, id, namePtr, context)
	if err != nil {
		return fmt.Errorf("inserting stakeholder: %w", err)
	}

	for _, org := range orgs {
		s.DB.Exec("INSERT IGNORE INTO stakeholder_orgs (tree, stakeholder_id, org) VALUES (?, ?, ?)",
			tree, id, strings.TrimSpace(org))
	}

	s.Commit(fmt.Sprintf("landscape(%s): add stakeholder %s", tree, id))
	return jsonOut(map[string]any{"tree": tree, "id": id, "context": context})
}

func cmdStakeholderList(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth stakeholder list <tree>")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree := args[0]
	rows, err := s.DB.Query("SELECT id, name, context FROM stakeholders WHERE tree = ? ORDER BY id", tree)
	if err != nil {
		return err
	}
	defer rows.Close()

	var stakeholders []map[string]any
	for rows.Next() {
		var id, context string
		var name *string
		rows.Scan(&id, &name, &context)
		sh := map[string]any{"id": id, "context": context}
		if name != nil {
			sh["name"] = *name
		}
		// Get orgs
		orgRows, _ := s.DB.Query("SELECT org FROM stakeholder_orgs WHERE tree = ? AND stakeholder_id = ?", tree, id)
		var orgs []string
		for orgRows.Next() {
			var org string
			orgRows.Scan(&org)
			orgs = append(orgs, org)
		}
		orgRows.Close()
		if len(orgs) > 0 {
			sh["orgs"] = orgs
		}
		stakeholders = append(stakeholders, sh)
	}
	return jsonOut(map[string]any{"tree": tree, "stakeholders": stakeholders})
}

func cmdDisposition(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth disposition <add|list|supersede> ...")
	}
	switch args[0] {
	case "add":
		return cmdDispositionAdd(args[1:])
	case "list":
		return cmdDispositionList(args[1:])
	case "supersede":
		return cmdDispositionSupersede(args[1:])
	default:
		return fmt.Errorf("unknown disposition subcommand: %s", args[0])
	}
}

func cmdDispositionAdd(args []string) error {
	if len(args) < 2 {
		return fmt.Errorf("usage: synth disposition add <tree/spec> <stakeholder> --topic '...' --stance cautious --confidence inferred [--preferred '...']")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree, spec, err := parseTreeSpec(args[0])
	if err != nil {
		return err
	}
	stakeholderID := args[1]

	var topic, stance, confidence, preferred string
	for i := 2; i < len(args)-1; i += 2 {
		switch args[i] {
		case "--topic":
			topic = args[i+1]
		case "--stance":
			stance = args[i+1]
		case "--confidence":
			confidence = args[i+1]
		case "--preferred":
			preferred = args[i+1]
		}
	}
	if topic == "" || stance == "" || confidence == "" {
		return fmt.Errorf("--topic, --stance, and --confidence are required")
	}

	// Get next ID
	var ids []string
	rows, _ := s.DB.Query("SELECT id FROM dispositions WHERE tree = ? AND spec = ?", tree, spec)
	for rows.Next() {
		var id string
		rows.Scan(&id)
		ids = append(ids, id)
	}
	rows.Close()
	newID := store.NextID("d", ids)

	var preferredPtr *string
	if preferred != "" {
		preferredPtr = &preferred
	}

	_, err = s.DB.Exec(
		"INSERT INTO dispositions (tree, spec, id, stakeholder_id, topic, stance, preferred_approach, confidence, valid_from) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)",
		tree, spec, newID, stakeholderID, topic, stance, preferredPtr, confidence, store.Today(),
	)
	if err != nil {
		return fmt.Errorf("inserting disposition: %w", err)
	}

	s.Commit(fmt.Sprintf("landscape(%s/%s): disposition %s -- %s is %s on %s", tree, spec, newID, stakeholderID, stance, topic))
	return jsonOut(map[string]any{
		"id": newID, "stakeholder": stakeholderID, "topic": topic,
		"stance": stance, "confidence": confidence,
	})
}

func cmdDispositionList(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth disposition list <tree/spec>")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree, spec, err := parseTreeSpec(args[0])
	if err != nil {
		return err
	}

	rows, err := s.DB.Query(
		"SELECT id, stakeholder_id, topic, stance, preferred_approach, confidence, valid_from, valid_until FROM dispositions WHERE tree = ? AND spec = ? ORDER BY valid_from DESC",
		tree, spec,
	)
	if err != nil {
		return err
	}
	defer rows.Close()

	var dispositions []map[string]any
	for rows.Next() {
		var id, stakeholder, topic, stance, confidence, validFrom string
		var preferred, validUntil *string
		rows.Scan(&id, &stakeholder, &topic, &stance, &preferred, &confidence, &validFrom, &validUntil)
		d := map[string]any{
			"id": id, "stakeholder": stakeholder, "topic": topic,
			"stance": stance, "confidence": confidence, "valid_from": validFrom,
			"current": validUntil == nil,
		}
		if preferred != nil {
			d["preferred_approach"] = *preferred
		}
		if validUntil != nil {
			d["valid_until"] = *validUntil
		}
		dispositions = append(dispositions, d)
	}
	return jsonOut(map[string]any{"tree": tree, "spec": spec, "dispositions": dispositions})
}

func cmdDispositionSupersede(args []string) error {
	if len(args) < 2 {
		return fmt.Errorf("usage: synth disposition supersede <tree/spec> <disposition-id> --new-stance supportive [--evidence s1] [--preferred '...']")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree, spec, err := parseTreeSpec(args[0])
	if err != nil {
		return err
	}
	oldID := args[1]

	var newStance, preferred, evidence string
	for i := 2; i < len(args)-1; i += 2 {
		switch args[i] {
		case "--new-stance":
			newStance = args[i+1]
		case "--preferred":
			preferred = args[i+1]
		case "--evidence":
			evidence = args[i+1]
		}
	}
	if newStance == "" {
		return fmt.Errorf("--new-stance is required")
	}

	// Read old disposition
	var stakeholder, topic, confidence string
	var oldPreferred *string
	err = s.DB.QueryRow(
		"SELECT stakeholder_id, topic, confidence, preferred_approach FROM dispositions WHERE tree = ? AND spec = ? AND id = ?",
		tree, spec, oldID,
	).Scan(&stakeholder, &topic, &confidence, &oldPreferred)
	if err != nil {
		return fmt.Errorf("disposition %s not found", oldID)
	}

	// Generate new ID
	var ids []string
	rows, _ := s.DB.Query("SELECT id FROM dispositions WHERE tree = ? AND spec = ?", tree, spec)
	for rows.Next() {
		var id string
		rows.Scan(&id)
		ids = append(ids, id)
	}
	rows.Close()
	newID := store.NextID("d", ids)

	today := store.Today()

	// Supersede old
	s.DB.Exec("UPDATE dispositions SET valid_until = ?, superseded_by = ? WHERE tree = ? AND spec = ? AND id = ?",
		today, newID, tree, spec, oldID)

	// Insert new
	var preferredPtr *string
	if preferred != "" {
		preferredPtr = &preferred
	} else {
		preferredPtr = oldPreferred
	}

	var evidencePtr *string
	if evidence != "" {
		evidencePtr = &evidence
	}

	s.DB.Exec(
		"INSERT INTO dispositions (tree, spec, id, stakeholder_id, topic, stance, preferred_approach, confidence, evidence, valid_from) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
		tree, spec, newID, stakeholder, topic, newStance, preferredPtr, confidence, evidencePtr, today,
	)

	commitMsg := fmt.Sprintf("landscape(%s/%s): supersede %s -> %s (%s now %s on %s)", tree, spec, oldID, newID, stakeholder, newStance, topic)
	if evidence != "" {
		commitMsg += fmt.Sprintf(" [evidence: %s]", evidence)
	}
	s.Commit(commitMsg)

	out := map[string]any{
		"old_id": oldID, "new_id": newID, "stakeholder": stakeholder,
		"topic": topic, "old_stance": "superseded", "new_stance": newStance,
	}
	if evidence != "" {
		out["evidence_signal"] = evidence
	}
	return jsonOut(out)
}

func cmdSignal(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth signal <record|list> ...")
	}
	switch args[0] {
	case "record":
		return cmdSignalRecord(args[1:])
	case "list":
		return cmdSignalList(args[1:])
	default:
		return fmt.Errorf("unknown signal subcommand: %s", args[0])
	}
}

func cmdSignalRecord(args []string) error {
	if len(args) < 2 {
		return fmt.Errorf("usage: synth signal record <tree/spec> <stakeholder> --source 'url' --type pr_comment --content '...' [--date YYYY-MM-DD] [--our-action '...'] [--interpretation '...']")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree, spec, err := parseTreeSpec(args[0])
	if err != nil {
		return err
	}
	stakeholderID := args[1]

	var source, sourceType, content, date, ourAction, interpretation string
	for i := 2; i < len(args)-1; i += 2 {
		switch args[i] {
		case "--source":
			source = args[i+1]
		case "--type":
			sourceType = args[i+1]
		case "--content":
			content = args[i+1]
		case "--date":
			date = args[i+1]
		case "--our-action":
			ourAction = args[i+1]
		case "--interpretation":
			interpretation = args[i+1]
		}
	}
	if source == "" || sourceType == "" || content == "" {
		return fmt.Errorf("--source, --type, and --content are required")
	}
	if date == "" {
		date = store.Today()
	}

	var ids []string
	rows, _ := s.DB.Query("SELECT id FROM signals WHERE tree = ? AND spec = ?", tree, spec)
	for rows.Next() {
		var id string
		rows.Scan(&id)
		ids = append(ids, id)
	}
	rows.Close()
	newID := store.NextID("s", ids)

	var ourActionPtr, interpPtr *string
	if ourAction != "" {
		ourActionPtr = &ourAction
	}
	if interpretation != "" {
		interpPtr = &interpretation
	}

	_, err = s.DB.Exec(
		"INSERT INTO signals (tree, spec, id, stakeholder_id, date, recorded_date, source, source_type, content, our_action, interpretation) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)",
		tree, spec, newID, stakeholderID, date, store.Today(), source, sourceType, content, ourActionPtr, interpPtr,
	)
	if err != nil {
		return fmt.Errorf("inserting signal: %w", err)
	}

	s.Commit(fmt.Sprintf("landscape(%s/%s): signal %s from %s", tree, spec, newID, stakeholderID))
	return jsonOut(map[string]any{
		"id": newID, "stakeholder": stakeholderID, "date": date,
		"source_type": sourceType, "recorded_date": store.Today(),
	})
}

func cmdSignalList(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth signal list <tree/spec>")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree, spec, err := parseTreeSpec(args[0])
	if err != nil {
		return err
	}

	rows, err := s.DB.Query(
		"SELECT id, stakeholder_id, date, recorded_date, source, source_type, content, our_action, interpretation FROM signals WHERE tree = ? AND spec = ? ORDER BY date DESC",
		tree, spec,
	)
	if err != nil {
		return err
	}
	defer rows.Close()

	var signals []map[string]any
	for rows.Next() {
		var id, stakeholder, date, recordedDate, source, sourceType, content string
		var ourAction, interpretation *string
		rows.Scan(&id, &stakeholder, &date, &recordedDate, &source, &sourceType, &content, &ourAction, &interpretation)
		sig := map[string]any{
			"id": id, "stakeholder": stakeholder, "date": date,
			"recorded_date": recordedDate, "source": source,
			"source_type": sourceType, "content": content,
		}
		if ourAction != nil {
			sig["our_action"] = *ourAction
		}
		if interpretation != nil {
			sig["interpretation"] = *interpretation
		}
		signals = append(signals, sig)
	}
	return jsonOut(map[string]any{"tree": tree, "spec": spec, "signals": signals})
}

func cmdLandscape(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth landscape show <tree/spec>")
	}
	if args[0] != "show" {
		return fmt.Errorf("unknown landscape subcommand: %s (did you mean 'show'?)", args[0])
	}
	return cmdLandscapeShow(args[1:])
}

func cmdLandscapeShow(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth landscape show <tree/spec>")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	tree, spec, err := parseTreeSpec(args[0])
	if err != nil {
		return err
	}

	result := map[string]any{"tree": tree, "spec": spec}

	// Current dispositions
	rows, _ := s.DB.Query(
		"SELECT d.id, d.stakeholder_id, s.context, d.topic, d.stance, d.preferred_approach, d.confidence, d.valid_from FROM dispositions d JOIN stakeholders s ON d.tree = s.tree AND d.stakeholder_id = s.id WHERE d.tree = ? AND d.spec = ? AND d.valid_until IS NULL ORDER BY d.stakeholder_id",
		tree, spec,
	)
	dispositions := make([]map[string]any, 0)
	for rows.Next() {
		var id, stakeholder, context, topic, stance, confidence, validFrom string
		var preferred *string
		rows.Scan(&id, &stakeholder, &context, &topic, &stance, &preferred, &confidence, &validFrom)
		d := map[string]any{
			"id": id, "stakeholder": stakeholder, "stakeholder_context": context,
			"topic": topic, "stance": stance, "confidence": confidence, "valid_from": validFrom,
		}
		if preferred != nil {
			d["preferred_approach"] = *preferred
		}
		dispositions = append(dispositions, d)
	}
	rows.Close()
	result["dispositions"] = dispositions

	// Signals
	rows, _ = s.DB.Query(
		"SELECT sig.id, sig.stakeholder_id, sh.context, sig.date, sig.source, sig.source_type, sig.content FROM signals sig JOIN stakeholders sh ON sig.tree = sh.tree AND sig.stakeholder_id = sh.id WHERE sig.tree = ? AND sig.spec = ? ORDER BY sig.date DESC LIMIT 20",
		tree, spec,
	)
	signals := make([]map[string]any, 0)
	for rows.Next() {
		var id, stakeholder, context, date, source, sourceType, content string
		rows.Scan(&id, &stakeholder, &context, &date, &source, &sourceType, &content)
		signals = append(signals, map[string]any{
			"id": id, "stakeholder": stakeholder, "stakeholder_context": context,
			"date": date, "source": source, "source_type": sourceType, "content": content,
		})
	}
	rows.Close()
	result["signals"] = signals

	// Influences
	rows, _ = s.DB.Query(
		"SELECT i.stakeholder_id, s.context, i.task_id, i.role FROM influences i JOIN stakeholders s ON i.tree = s.tree AND i.stakeholder_id = s.id WHERE i.tree = ? AND i.spec = ?",
		tree, spec,
	)
	influences := make([]map[string]any, 0)
	for rows.Next() {
		var stakeholder, context, taskID, role string
		rows.Scan(&stakeholder, &context, &taskID, &role)
		influences = append(influences, map[string]any{
			"stakeholder": stakeholder, "context": context, "task": taskID, "role": role,
		})
	}
	rows.Close()
	result["influences"] = influences

	// Directions affecting this spec
	rows, _ = s.DB.Query(`
		SELECT d.id, d.project, d.topic, d.status, d.impact, di.description
		FROM directions d
		JOIN direction_impacts di ON d.tree = di.tree AND d.id = di.direction_id
		WHERE di.affected_tree = ? AND di.affected_spec = ? AND d.valid_until IS NULL
	`, tree, spec)
	directions := make([]map[string]any, 0)
	for rows.Next() {
		var id, project, topic, status, impact, desc string
		rows.Scan(&id, &project, &topic, &status, &impact, &desc)
		directions = append(directions, map[string]any{
			"id": id, "project": project, "topic": topic,
			"status": status, "impact": impact, "impact_on_spec": desc,
		})
	}
	rows.Close()
	result["directions"] = directions

	return jsonOut(result)
}

func cmdStance(args []string) error {
	if len(args) < 1 {
		return fmt.Errorf("usage: synth stance <stakeholder> [topic]")
	}
	s, err := discoverStore()
	if err != nil {
		return err
	}
	defer s.Close()

	stakeholderID := args[0]
	var topic string
	if len(args) > 1 {
		topic = args[1]
	}

	var rows interface{ Next() bool; Scan(...any) error; Close() error }
	if topic != "" {
		// Full history for this person + topic
		r, err := s.DB.Query(
			"SELECT tree, spec, id, stance, preferred_approach, confidence, valid_from, valid_until FROM dispositions WHERE stakeholder_id = ? AND topic = ? ORDER BY valid_from DESC",
			stakeholderID, topic,
		)
		if err != nil {
			return err
		}
		rows = r
	} else {
		// Current dispositions across all specs
		r, err := s.DB.Query(
			"SELECT tree, spec, id, topic, stance, preferred_approach, confidence, valid_from FROM dispositions WHERE stakeholder_id = ? AND valid_until IS NULL ORDER BY valid_from DESC",
			stakeholderID,
		)
		if err != nil {
			return err
		}
		rows = r
	}
	defer rows.Close()

	var dispositions []map[string]any
	for rows.Next() {
		if topic != "" {
			var tree, spec, id, stance, confidence, validFrom string
			var preferred, validUntil *string
			rows.Scan(&tree, &spec, &id, &stance, &preferred, &confidence, &validFrom, &validUntil)
			d := map[string]any{
				"tree": tree, "spec": spec, "id": id, "stance": stance,
				"confidence": confidence, "valid_from": validFrom,
				"current": validUntil == nil,
			}
			if preferred != nil {
				d["preferred_approach"] = *preferred
			}
			if validUntil != nil {
				d["valid_until"] = *validUntil
			}
			dispositions = append(dispositions, d)
		} else {
			var tree, spec, id, dtopic, stance, confidence, validFrom string
			var preferred *string
			rows.Scan(&tree, &spec, &id, &dtopic, &stance, &preferred, &confidence, &validFrom)
			d := map[string]any{
				"tree": tree, "spec": spec, "id": id, "topic": dtopic,
				"stance": stance, "confidence": confidence, "valid_from": validFrom,
			}
			if preferred != nil {
				d["preferred_approach"] = *preferred
			}
			dispositions = append(dispositions, d)
		}
	}

	result := map[string]any{"stakeholder": stakeholderID, "dispositions": dispositions}
	if topic != "" {
		result["topic"] = topic
		result["mode"] = "history"
	} else {
		result["mode"] = "current"
	}
	return jsonOut(result)
}

File: src/cmd_landscape.rs
//! Stakeholder, disposition, signal, and stance commands.

use anyhow::Result;
use serde_json::json;

use crate::cli::{DispositionCmd, SignalCmd, StakeholderCmd};
use crate::store::{json_out, parse_tree_spec, Store};

// --- Stakeholder ---

pub fn run_stakeholder(cmd: &StakeholderCmd, session: &Option<String>) -> Result<()> {
    match cmd {
        StakeholderCmd::Add {
            tree,
            id,
            context,
            name,
            orgs,
        } => cmd_stakeholder_add(tree, id, context, name.as_deref(), orgs, session),
        StakeholderCmd::List { tree } => cmd_stakeholder_list(tree, session),
    }
}

fn cmd_stakeholder_add(
    tree: &str,
    id: &str,
    context: &str,
    name: Option<&str>,
    orgs: &[String],
    session: &Option<String>,
) -> Result<()> {
    let store = Store::discover_for(session)?;
    store.conn.execute(
        "INSERT INTO stakeholders (tree, id, name, context) VALUES (?1, ?2, ?3, ?4)",
        rusqlite::params![tree, id, name, context],
    )?;
    for org in orgs {
        if org.is_empty() {
            continue;
        }
        store.conn.execute(
            "INSERT INTO stakeholder_orgs (tree, stakeholder_id, org) VALUES (?1, ?2, ?3)",
            rusqlite::params![tree, id, org],
        )?;
    }
    json_out(&json!({"tree": tree, "id": id, "context": context}))
}

fn cmd_stakeholder_list(tree: &str, session: &Option<String>) -> Result<()> {
    let store = Store::discover_for(session)?;
    let mut stmt = store
        .conn
        .prepare("SELECT id, name, context FROM stakeholders WHERE tree = ?1 ORDER BY id")?;
    let stakeholders: Vec<serde_json::Value> = stmt
        .query_map([tree], |row| {
            Ok(json!({
                "id": row.get::<_, String>(0)?,
                "name": row.get::<_, Option<String>>(1)?,
                "context": row.get::<_, String>(2)?,
            }))
        })?
        .collect::<std::result::Result<Vec<_>, rusqlite::Error>>()?;
    json_out(&json!({"tree": tree, "stakeholders": stakeholders}))
}

// --- Disposition ---

pub fn run_disposition(cmd: &DispositionCmd, session: &Option<String>) -> Result<()> {
    match cmd {
        DispositionCmd::Add {
            tree_spec,
            stakeholder,
            topic,
            stance,
            confidence,
            preferred,
            detail,
        } => {
            let (tree, spec) = parse_tree_spec(tree_spec)?;
            cmd_disposition_add(
                tree,
                spec,
                stakeholder,
                topic,
                stance,
                confidence,
                preferred.as_deref(),
                detail.as_deref(),
                session,
            )
        }
        DispositionCmd::List { tree_spec } => {
            let (tree, spec) = parse_tree_spec(tree_spec)?;
            cmd_disposition_list(tree, spec, session)
        }
        DispositionCmd::Supersede {
            tree_spec,
            old_id,
            stance,
            confidence,
            preferred,
            detail,
        } => {
            let (tree, spec) = parse_tree_spec(tree_spec)?;
            cmd_disposition_supersede(
                tree,
                spec,
                old_id,
                stance,
                confidence,
                preferred.as_deref(),
                detail.as_deref(),
                session,
            )
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn cmd_disposition_add(
    tree: &str,
    spec: &str,
    stakeholder: &str,
    topic: &str,
    stance: &str,
    confidence: &str,
    preferred: Option<&str>,
    detail: Option<&str>,
    session: &Option<String>,
) -> Result<()> {
    let store = Store::discover_for(session)?;
    let id = store.next_id("dispositions", tree, spec, "disp")?;
    let today = Store::today();
    store.conn.execute(
        "INSERT INTO dispositions (tree, spec, id, stakeholder_id, topic, stance, preferred_approach, detail, confidence, valid_from) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10)",
        rusqlite::params![tree, spec, id, stakeholder, topic, stance, preferred, detail, confidence, today],
    )?;
    json_out(&json!({
        "id": id,
        "stakeholder": stakeholder,
        "topic": topic,
        "stance": stance,
        "confidence": confidence,
    }))
}

fn cmd_disposition_list(tree: &str, spec: &str, session: &Option<String>) -> Result<()> {
    let store = Store::discover_for(session)?;
    let mut stmt = store.conn.prepare(
        "SELECT id, stakeholder_id, topic, stance, preferred_approach, detail, confidence, valid_from, valid_until, superseded_by FROM dispositions WHERE tree = ?1 AND spec = ?2 ORDER BY valid_from DESC",
    )?;
    let dispositions: Vec<serde_json::Value> = stmt
        .query_map(rusqlite::params![tree, spec], |row| {
            Ok(json!({
                "id": row.get::<_, String>(0)?,
                "stakeholder": row.get::<_, String>(1)?,
                "topic": row.get::<_, String>(2)?,
                "stance": row.get::<_, String>(3)?,
                "preferred_approach": row.get::<_, Option<String>>(4)?,
                "detail": row.get::<_, Option<String>>(5)?,
                "confidence": row.get::<_, String>(6)?,
                "valid_from": row.get::<_, String>(7)?,
                "valid_until": row.get::<_, Option<String>>(8)?,
                "superseded_by": row.get::<_, Option<String>>(9)?,
            }))
        })?
        .collect::<std::result::Result<Vec<_>, rusqlite::Error>>()?;
    json_out(&json!({"tree": tree, "spec": spec, "dispositions": dispositions}))
}

#[allow(clippy::too_many_arguments)]
fn cmd_disposition_supersede(
    tree: &str,
    spec: &str,
    old_id: &str,
    stance: &str,
    confidence: &str,
    preferred: Option<&str>,
    detail: Option<&str>,
    session: &Option<String>,
) -> Result<()> {
    let store = Store::discover_for(session)?;

    // Get the old disposition's stakeholder and topic
    let (stakeholder, topic): (String, String) = store
        .conn
        .query_row(
            "SELECT stakeholder_id, topic FROM dispositions WHERE tree = ?1 AND spec = ?2 AND id = ?3",
            rusqlite::params![tree, spec, old_id],
            |row| Ok((row.get(0)?, row.get(1)?)),
        )
        .map_err(|_| anyhow::anyhow!("disposition not found: {tree}/{spec}/{old_id}"))?;

    let new_id = store.next_id("dispositions", tree, spec, "disp")?;
    let today = Store::today();

    // Wrap supersession in a transaction (two rows must update atomically).
    store.conn.execute("BEGIN IMMEDIATE", [])?;

    let result = (|| -> Result<()> {
        store.conn.execute(
            "INSERT INTO dispositions (tree, spec, id, stakeholder_id, topic, stance, preferred_approach, detail, confidence, valid_from) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10)",
            rusqlite::params![tree, spec, new_id, stakeholder, topic, stance, preferred, detail, confidence, today],
        )?;
        store.conn.execute(
            "UPDATE dispositions SET valid_until = ?1, superseded_by = ?2 WHERE tree = ?3 AND spec = ?4 AND id = ?5",
            rusqlite::params![today, new_id, tree, spec, old_id],
        )?;
        Ok(())
    })();

    match result {
        Ok(()) => store.conn.execute("COMMIT", [])?,
        Err(e) => {
            store.conn.execute("ROLLBACK", []).ok();
            return Err(e);
        }
    };

    json_out(&json!({
        "old_id": old_id,
        "new_id": new_id,
        "stakeholder": stakeholder,
        "topic": topic,
        "stance": stance,
    }))
}

// --- Signal ---

pub fn run_signal(cmd: &SignalCmd, session: &Option<String>) -> Result<()> {
    match cmd {
        SignalCmd::Add {
            tree_spec,
            stakeholder,
            source,
            source_type,
            content,
            interpretation,
            our_action,
            date,
        } => {
            let (tree, spec) = parse_tree_spec(tree_spec)?;
            cmd_signal_add(
                tree,
                spec,
                stakeholder,
                source,
                source_type,
                content,
                interpretation.as_deref(),
                our_action.as_deref(),
                date.as_deref(),
                session,
            )
        }
        SignalCmd::List { tree_spec } => {
            let (tree, spec) = parse_tree_spec(tree_spec)?;
            cmd_signal_list(tree, spec, session)
        }
    }
}

#[allow(clippy::too_many_arguments)]
fn cmd_signal_add(
    tree: &str,
    spec: &str,
    stakeholder: &str,
    source: &str,
    source_type: &str,
    content: &str,
    interpretation: Option<&str>,
    our_action: Option<&str>,
    date: Option<&str>,
    session: &Option<String>,
) -> Result<()> {
    let store = Store::discover_for(session)?;
    let id = store.next_id("signals", tree, spec, "sig")?;
    let event_date = date.unwrap_or(&Store::today()).to_string();
    let recorded = Store::today();
    store.conn.execute(
        "INSERT INTO signals (tree, spec, id, stakeholder_id, date, recorded_date, source, source_type, content, interpretation, our_action) VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11)",
        rusqlite::params![tree, spec, id, stakeholder, event_date, recorded, source, source_type, content, interpretation, our_action],
    )?;
    json_out(&json!({"id": id, "stakeholder": stakeholder, "source": source, "date": event_date}))
}

fn cmd_signal_list(tree: &str, spec: &str, session: &Option<String>) -> Result<()> {
    let store = Store::discover_for(session)?;
    let mut stmt = store.conn.prepare(
        "SELECT id, stakeholder_id, date, recorded_date, source, source_type, content, interpretation, our_action FROM signals WHERE tree = ?1 AND spec = ?2 ORDER BY date DESC",
    )?;
    let signals: Vec<serde_json::Value> = stmt
        .query_map(rusqlite::params![tree, spec], |row| {
            Ok(json!({
                "id": row.get::<_, String>(0)?,
                "stakeholder": row.get::<_, String>(1)?,
                "date": row.get::<_, String>(2)?,
                "recorded_date": row.get::<_, String>(3)?,
                "source": row.get::<_, String>(4)?,
                "source_type": row.get::<_, String>(5)?,
                "content": row.get::<_, String>(6)?,
                "interpretation": row.get::<_, Option<String>>(7)?,
                "our_action": row.get::<_, Option<String>>(8)?,
            }))
        })?
        .collect::<std::result::Result<Vec<_>, rusqlite::Error>>()?;
    json_out(&json!({"tree": tree, "spec": spec, "signals": signals}))
}

// --- Stance query ---

pub fn cmd_stance(stakeholder: &str, topic: Option<&str>, session: &Option<String>) -> Result<()> {
    let store = Store::discover_for(session)?;

    let sql = if topic.is_some() {
        "SELECT d.tree, d.spec, d.id, d.topic, d.stance, d.preferred_approach, d.detail, d.confidence, d.valid_from
         FROM dispositions d
         JOIN stakeholders s ON d.tree = s.tree AND d.stakeholder_id = s.id
         WHERE s.id = ?1 AND d.topic LIKE ?2 AND d.valid_until IS NULL
         ORDER BY d.valid_from DESC"
    } else {
        "SELECT d.tree, d.spec, d.id, d.topic, d.stance, d.preferred_approach, d.detail, d.confidence, d.valid_from
         FROM dispositions d
         JOIN stakeholders s ON d.tree = s.tree AND d.stakeholder_id = s.id
         WHERE s.id = ?1 AND d.valid_until IS NULL
         ORDER BY d.valid_from DESC"
    };

    let topic_pattern = topic.map(|t| format!("%{t}%")).unwrap_or_default();
    let params: Vec<&dyn rusqlite::types::ToSql> = if topic.is_some() {
        vec![&stakeholder as &dyn rusqlite::types::ToSql, &topic_pattern]
    } else {
        vec![&stakeholder as &dyn rusqlite::types::ToSql]
    };

    let mut stmt = store.conn.prepare(sql)?;
    let dispositions: Vec<serde_json::Value> = stmt
        .query_map(params.as_slice(), |row| {
            Ok(json!({
                "tree": row.get::<_, String>(0)?,
                "spec": row.get::<_, String>(1)?,
                "id": row.get::<_, String>(2)?,
                "topic": row.get::<_, String>(3)?,
                "stance": row.get::<_, String>(4)?,
                "preferred_approach": row.get::<_, Option<String>>(5)?,
                "detail": row.get::<_, Option<String>>(6)?,
                "confidence": row.get::<_, String>(7)?,
                "valid_from": row.get::<_, String>(8)?,
            }))
        })?
        .collect::<std::result::Result<Vec<_>, rusqlite::Error>>()?;

    json_out(&json!({"stakeholder": stakeholder, "dispositions": dispositions}))
}
